reactor-netty-http/src/main/java/reactor/netty/http/client/AbstractHttpClientMetricsHandler.java

```diff
@@ -76,64 +80,88 @@ protected AbstractHttpClientMetricsHandler(AbstractHttpClientMetricsHandler copy
 	@Override
 	@SuppressWarnings("FutureReturnValueIgnored")
 	public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-		if (msg instanceof HttpRequest) {
-			method = ((HttpRequest) msg).method().name();
-
-			ChannelOperations<?, ?> channelOps = ChannelOperations.get(ctx.channel());
-			if (channelOps instanceof HttpClientOperations) {
-				HttpClientOperations ops = (HttpClientOperations) channelOps;
-				path = uriTagValue == null ? ops.path : uriTagValue.apply(ops.path);
-				contextView = ops.currentContextView();
+		try {
+			if (msg instanceof HttpRequest) {
+				extractDetailsFromHttpRequest(ctx, (HttpRequest) msg);
 			}
 
-			dataSentTime = System.nanoTime();
-		}
-
-		if (msg instanceof ByteBufHolder) {
-			dataSent += ((ByteBufHolder) msg).content().readableBytes();
-		}
-		else if (msg instanceof ByteBuf) {
-			dataSent += ((ByteBuf) msg).readableBytes();
-		}
-
-		if (msg instanceof LastHttpContent) {
-			SocketAddress address = ctx.channel().remoteAddress();
-			promise.addListener(future -> recordWrite(address));
+			dataSent += extractProcessedDataFromBuffer(msg);
+
+			if (msg instanceof LastHttpContent) {
+				SocketAddress address = ctx.channel().remoteAddress();
+				promise.addListener(future -> {
+					try {
+						recordWrite(address);
+					} catch (RuntimeException e) {
+						log.warn("Exception caught while recording metrics.", e);
+						// Allow request-response exchange to continue, unaffected by metrics problem
+					}
+				});
+			}
+		} catch (RuntimeException e) {
+			log.warn("Exception caught while recording metrics.", e);
+			// Allow request-response exchange to continue, unaffected by metrics problem
 		}
 		//"FutureReturnValueIgnored" this is deliberate
 		ctx.write(msg, promise);
 	}
 
 	@Override
 	public void channelRead(ChannelHandlerContext ctx, Object msg) {
-		if (msg instanceof HttpResponse) {
-			status = ((HttpResponse) msg).status().codeAsText().toString();
+		try {
+			if (msg instanceof HttpResponse) {
+				status = ((HttpResponse) msg).status().codeAsText().toString();
 
-			dataReceivedTime = System.nanoTime();
-		}
+				dataReceivedTime = System.nanoTime();
+			}
 
-		if (msg instanceof ByteBufHolder) {
-			dataReceived += ((ByteBufHolder) msg).content().readableBytes();
-		}
-		else if (msg instanceof ByteBuf) {
-			dataReceived += ((ByteBuf) msg).readableBytes();
-		}
+			dataReceived += extractProcessedDataFromBuffer(msg);
 
-		if (msg instanceof LastHttpContent) {
-			recordRead(ctx.channel().remoteAddress());
-			reset();
+			if (msg instanceof LastHttpContent) {
+				recordRead(ctx.channel().remoteAddress());
+				reset();
+			}
+		} catch (RuntimeException e) {
+			log.warn("Exception caught while recording metrics.", e);
+			// Allow request-response exchange to continue, unaffected by metrics problem
```

Comment:
Do we need this `try/catch` for the whole block or it is more appropriate to have it only for the recording? Is it possible to have a `RuntimeException` not related to metrics?

Reply:
This is a good question. I'm not very familiar with the objects used here. `status = ((HttpResponse) msg).status().codeAsText().toString();` does look like it could result in a `NullPointerException` also other invocation don't clearly identify as safe to me. All of the enclosed code is only contributing to metrics recording and should IMHO not terminate data processing. I'd be glad to reduce the scope if it's certain that there will be no issues with those calls.